packages/asset-uploader/src/targets/s3/upload.ts

```diff
@@ -34,32 +38,36 @@ export const uploadToS3 = async ({
   projectId: string;
   maxSize: number;
 }): Promise<Array<Asset>> => {
-  const uploadHandler = createUploadHandler(MAX_FILES_PER_REQUEST);
-
-  const formData = await unstableCreateFileUploadHandler(
-    request,
-    unstableComposeUploadHandlers(
-      (file: UploadHandlerPart) =>
-        uploadHandler({
-          file,
-          maxSize,
-        }),
-      uuidHandler
-    )
-  );
-
-  const imagesFormData = formData.getAll("image") as Array<string>;
-  const fontsFormData = formData.getAll("font") as Array<string>;
-  const ids = Ids.parse(formData.getAll(idsFormDataFieldName));
-
-  const assetsData = [...imagesFormData, ...fontsFormData]
-    .slice(0, MAX_FILES_PER_REQUEST)
-    .map((dataString, i) => {
-      // @todo validate with zod
-      return { ...JSON.parse(dataString), id: ids[i] };
-    });
-
-  return await createMany(projectId, assetsData);
+  const asset = await createAssetWithLimit(projectId, async () => {
+    const uploadHandler = createUploadHandler(MAX_FILES_PER_REQUEST);
+
+    const formData = await unstableCreateFileUploadHandler(
+      request,
+      unstableComposeUploadHandlers(
+        (file: UploadHandlerPart) =>
+          uploadHandler({
+            file,
+            maxSize,
+          }),
+        uuidHandler
+      )
+    );
+
+    const imagesFormData = formData.getAll("image") as Array<string>;
+    const fontsFormData = formData.getAll("font") as Array<string>;
+    const ids = Ids.parse(formData.getAll(idsFormDataFieldName));
+
+    const assetsData = [...imagesFormData, ...fontsFormData]
+      .slice(0, MAX_FILES_PER_REQUEST)
+      .map((dataString, i) => {
+        // @todo validate with zod
+        return { ...JSON.parse(dataString), id: ids[i] };
```

Comment:
looks a little suspicious that ids index would correspond to assets

Reply:
Yep, the only thing that we restrict arrays to `size 1` and we know that we pass the same arrays length=1 from the client.
Looks like the only way this not being suspicious is to pass Blob instead of the file from the client and parse id and file content from the Blob.

Reply:
or run some validation before using the index

Reply:
Added zod validation of the whole final object.